_posts/2019-04-24-istio-service-mesh-s2s.md

```diff
@@ -0,0 +1,431 @@
+* [Show me the code](#show-me-the-code)
+* [Conclusion](#conclusion)
+
+## What is Istio?
+
+Istio is a service mesh created by Google, Lyft and IBM. 
+It aims to simplify some security and management aspects of a microservices software architecture.
+More information on Istio and its feature can be found in its [docs](https://istio.io/docs/){:target="_blank" rel="noopener noreferrer"}.
+In this blogpost we will highlight one of the key security features of Istio: service to service authentication and authorization.
+For the sake of simplicity, this post will focus on an Istio setup in Kubernetes.
+
+In a microservices architecture, managing access to services can be a challenging operation. 
+For end-user facing services, JWTs are often used to authenicate a request.
+This token can be generated by an identity provider. 
+A client can provide it as means to make itself known to a service.
+For users, these tokens can be generated based on information that the end-user provides. 
+In most cases this information is a username and password, with some additional 2FA if possible.
+
+When authenticating services, providing an identity is a bit harder.
+One option is to use the same credential flow for services as for end-users.
+A service would provide a username and password and receive a (temporarily valid) token to access a service.
+Unfortunaly this approach has some drawbacks.
+The first main issue is that services often need more broader access to services than end-users.
+This makes the credentials they hold very powerful, which makes them dangerous to manage. 
+Secondly, the JWT is only valid for a predefined amount of time. 
+For a normal user, this is fine, since they can easily reauthenticate once the token expires.
+In services, reauthentication mechanisms tend to be a bit harder to set up and often result in hard to debug issues. 
+
+Introducing service to service authentication through mutual trusted certificates. 
+This approach uses public certificate and private key pairs to generate trust between services.
+When a service is started, a service specific certificate is generated and provided to the service.
+The service exposes this certificate and connecting services (clients) can validate it.
+Because certificates can have a longer life span, this is easier to manage, compared to the JWT setup. 
+Setting up certificates, adopting the services to use them and rotating them over time, can be very challenging.
+Every services needs to expose a secure endpoint and serve the correct certificate.
+This is a non-trivial step and can, again, introduce very hard to debug issues.
+
+This is exactly where Istio pops up in the game.
+It makes using certificates for service to service encryption and authentication easy.
+Especially when using the automatic proxy injection, it's almost as easy as flipping a switch.
+
+There are three main components responsible for making this possible in Istio: Citadel, Pilot and Mixer. 
```

Comment:
With Galley added, does that make it have four main components now?

Reply:
@TYsewyn I will include Envoy. I left it out because it isn't part of the Istio project. The mTLS stuff is discussed further down in the post :)